### PR TITLE
Pie-chart hover events

### PR DESCRIPTION
**Pull request by mdtusz**
The `fx.hover` method needed a bit of a tweak because pie gd's don't seem to have a `_fullLayout.plots`.

### CODE REVIEW

File: src/plots/cartesian/graph_interact.js [hunk 1]
```diff
@@ -309,6 +309,13 @@ fx.unhover = function (gd, evt, subplot) {
 // The actual implementation is here:
 
 function hover(gd, evt, subplot){
+    if(gd._fullLayout._hasPie){
```
Comment (etpinard) [inline src/plots/cartesian/graph_interact.js]: @mdtusz you're right.

The concept of (sub)plots doesn't extend to pies. Pies are considered traces that have thier own domain and no corresponding axes.

What does `evt` contain here?

File: src/traces/pie/index.js [hunk 2]
```diff
@@ -363,6 +363,8 @@ pie.plot = function(gd, cdpie) {
                         }
                     );
 
+                    Plotly.Fx.hover(gd, evt, 'pie');
```
Comment (etpinard) [inline src/traces/pie/index.js]: maybe 

```js
Plotly.Fx.hover(gd, evt, null);
```

would be more explicit.

Comment (mdtusz) [inline src/traces/pie/index.js]: Good call. The `pie` isn't really necessary here.

File: src/plots/cartesian/graph_interact.js (hunk 1 above)
Comment (mdtusz) [inline src/plots/cartesian/graph_interact.js]: `evt` contains the data from the hover event - in this case, information on the slice that was hovered over. The `plotly_hover` event fires and gets passed data in the same format as the click events.

Comment (etpinard) [inline src/plots/cartesian/graph_interact.js]: Right, but `Fx.hover` for pie traces should emit that same [data](https://github.com/plotly/plotly.js/blob/master/src/plots/cartesian/graph_interact.js#L573) as for non-pie traces minus the info about the axes of course.

Comment (mdtusz) [inline src/plots/cartesian/graph_interact.js]: The contents of `[evt]` mimic the same "format" as other traces hover events `points` value (as much as we can with a pie chart) but `xvals/yvals` and `xaxes/yaxes` are out of context here. The `xvals/yvals` are the mouse hover position relative to the axes scaling and values, so without any axes, these are meaningless.

Comment (etpinard) [inline src/plots/cartesian/graph_interact.js]: Great. Thanks for the clarification. :beers:

**Comment by etpinard** [comment]
@mdtusz 

Looks good.

If you're up for it, it'd be nice to add a test in https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/hover_label_test.js for pie hover labels.

**Comment by mdtusz** [comment]
Not sure why tests are failing... they all pass locally. Can you try pulling the branch down and testing it locally @etpinard?

File: test/jasmine/assets/mouse_event.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+module.exports = function (type, x, y, opts) {
+    var el = document.elementFromPoint(x,y);
+    var options = opts || { bubbles: true };
+    var ev = new window.MouseEvent(type, options);
+    el.dispatchEvent(ev);
```
Comment (etpinard) [inline test/jasmine/assets/mouse_event.js]: great. :beers:

**Comment by etpinard** [comment]
@mdtusz here's why:

image


The local tests run on Chrome by default whereas the CircleCi tests run on FF.

Looks like FF is more precise than Chrome here.

**Comment by mdtusz** [comment]
Ahh lovely. Ok. I'll have to make the test expectations a lil bit more/less specific then. Good catch though. I scanned the values but my eyes must have swapped the 5 and 6!

**Comment by etpinard** [comment]
@mdtusz maybe use Jasmine's `toBeCloseTo` (ref http://jsfiddle.net/lucassus/4DrrW/light/)

File: src/plots/cartesian/graph_interact.js (hunk 1 above)
Comment (etpinard) [inline src/plots/cartesian/graph_interact.js]: @mdtusz On second thought, a check on `_hasPie` is too coerced.

Say we have a pie and bar trace on the same graph:

```js
var gd = Tabs.fresh();

Plotly.plot(gd, 
    [
        {
            type: 'pie',
            labels: ['a', 'b', 'c'],
            values: [1, 2, 3],
            domain: {
                x: [0, 0.45]
            }
        },
        {
            type: 'bar',
            x: ['A', 'B', 'C'],
            y: [1, 2, 3],
        }
    
    ],
    {
        title: 'Pie vs bar',
        xaxis: { domain: [0.55, 1] },
        showlegend: false
    }
);

gd.on('plotly_hover', function(data) {
    var pt = data.points[0]
    console.log(pt);
});
```

then hovering over the bars will result in an un-formatted hover data.

So, we should use the `subplot` argument and make the if-clause: `if(subplot === 'pie')`.

File: test/jasmine/tests/hover_pie_test.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                futureData.push(data);
+            });
+
+            mouseEvent('mouseover', 180, 140);
+            setTimeout(function () {
+                mouseEvent('mouseover', 240, 200);
+                expect(count).toEqual(2);
+                expect(futureData[0]).not.toEqual(futureData[1])
+                done();
+            }, 100);
```
Comment (etpinard) [inline test/jasmine/tests/hover_pie_test.js]: Nice. Great test!

File: test/jasmine/tests/hover_pie_test.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            });
+
+            mouseEvent('mouseover', width / 2, height / 2);
+            expect(futureData.points.length).toEqual(1);
+            expect(Object.keys(futureData.points[0])).toEqual([
+                    'v', 'label', 'color', 'i', 'hidden',
+                    'text', 'px1', 'pxmid', 'midangle',
+                    'px0', 'largeArc', 'cxFinal', 'cyFinal'
+                ]);
+            expect(futureData.points[0].i).toEqual(3);
```
Comment (etpinard) [inline test/jasmine/tests/hover_pie_test.js]: yep, that's all we need to test here.

File: src/plots/cartesian/graph_interact.js (hunk 1 above)
Comment (mdtusz) [inline src/plots/cartesian/graph_interact.js]: Good catch.

**Comment by mdtusz** [comment]
@etpinard anything else we want to add to this pr?

**Comment by etpinard** [comment]
:dancer:

**Pull request merged**